File: JavaForest/src/cgNode/expNode.java
package cgNode;

public class expNode extends cgNode {
    public expNode(cgNode parent, boolean lr, cgNode left, cgNode right){
        super(parent, lr, left, right, 0);
    }

    @Override
    public cgNode copy(cgNode parent){
        cgNode newnode = new expNode(parent, _left_or_right, null, null);
        newnode._left = _left.copy(newnode);
        newnode._right = _right.copy(newnode);
        return newnode;
    }

    @Override
    public double compute(double n1, double n2) {
        return Math.pow(n1, n2);
    }

    @Override
    public cgNode diffor() {

        return this;
    }

    @Override
    public cgNode trim() {
        _left = _left.trim();
        _right = _right.trim();

        if (_left.is_n_node(0)){
            _left = new constNode(this, false, 0);
        }else if (_right.is_n_node(0)){
            return change_this(new constNode(_parent, _left_or_right, 1));
        }else if (_right.is_n_node(1)){
            return change_this(_left);
        }else if (_left.is_n_node(1)){
            return change_this(new constNode(_parent, _left_or_right, 1));
        }
        return this;
    }
}

File: cppForest/cgNode/expNode.cpp
//
// Created by a on 2019/5/14.
//

#include "expNode.h"
#include <cmath>

expNode::expNode(cgNode *parent, bool lr, cgNode *left, cgNode *right) : cgNode(parent, lr, left, right, 0) {}

cgNode *expNode::clone(cgNode *parent) {
    cgNode* p = new expNode(parent, _lr, nullptr, nullptr);
    p->_left = _left->clone(p);
    p->_right = _right->clone(p);
    return p;
}

double expNode::compute(double n1, double n2) {
    return pow(n1, n2);
}

cgNode *expNode::trim() {
    return this;
}

cgNode *expNode::diff() {
    return this;
}
